contracts/libraries/OperatorLib.sol

```diff
@@ -30,67 +33,208 @@ library OperatorLib {
+                    if (whitelistedAddress == address(0)) {
+                        // msg.sender is not whitelisted via bitmap or legacy whitelist/whitelisting contract
+                        revert ISSVNetworkCore.CallerNotWhitelisted(operatorId);
+                    }
+                    // Legacy address whitelists (EOAs or generic contracts)
+                    if (whitelistedAddress != msg.sender) {
+                        // Check if msg.sender is whitelisted via whitelisting contract
+                        if (!OperatorLib.isWhitelistingContract(whitelistedAddress)) {
+                            revert ISSVNetworkCore.InvalidWhitelistingContract(whitelistedAddress);
+                        }
+                        if (!ISSVWhitelistingContract(whitelistedAddress).isWhitelisted(msg.sender, operatorId)) {
+                            revert ISSVNetworkCore.CallerNotWhitelisted(operatorId);
+                        }
                     }
                 }
+            }
+
+            updateSnapshot(operator);
+            if ((operator.validatorCount += deltaValidatorCount) > sp.validatorsPerOperatorLimit) {
+                revert ISSVNetworkCore.ExceedValidatorLimit();
+            }
+
+            cumulativeFee += operator.fee;
+            cumulativeIndex += operator.snapshot.index;
+
+            s.operators[operatorId] = operator;
+
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    function updateClusterOperators(
+        uint64[] memory operatorIds,
+        bool increaseValidatorCount,
+        uint32 deltaValidatorCount,
+        StorageData storage s,
+        StorageProtocol storage sp
+    ) internal returns (uint64 cumulativeIndex, uint64 cumulativeFee) {
+        uint256 operatorsLength = operatorIds.length;
+
+        for (uint256 i; i < operatorsLength; ) {
+            uint64 operatorId = operatorIds[i];
+
+            ISSVNetworkCore.Operator storage operator = s.operators[operatorId];
 
-                updateSnapshot(operator);
-                if ((operator.validatorCount += deltaValidatorCount) > sp.validatorsPerOperatorLimit) {
+            if (operator.snapshot.block != 0) {
+                updateSnapshotSt(operator);
+                if (!increaseValidatorCount) {
+                    operator.validatorCount -= deltaValidatorCount;
+                } else if ((operator.validatorCount += deltaValidatorCount) > sp.validatorsPerOperatorLimit) {
                     revert ISSVNetworkCore.ExceedValidatorLimit();
                 }
 
                 cumulativeFee += operator.fee;
-                cumulativeIndex += operator.snapshot.index;
-
-                s.operators[operatorId] = operator;
             }
+            cumulativeIndex += operator.snapshot.index;
 
             unchecked {
                 ++i;
             }
         }
     }
+
+    function updateMultipleWhitelists(
```

Comment:
We dont do any checking of duplicate addresses or duplicate operator ids, i dont see an issue with not checking this just making sure theres nothing maybe with bitmap or something else im not missing

Reply:
@andrew-blox for `setOperatorMultipleWhitelists` and `removeOperatorMultipleWhitelists` we require the list of operators to be unique and ordered. The check is performed in `OperatorLib.generateBlockMasks()` function.
In `OperatorLib.updateMultipleWhitelists` we check for empty `operatorIds` and `whitelistAddresses`.

In `setOperatorsWhitelistingContract` we check for a valid whitelisting contract and implicitly for empty address, also empty `operatorIds`. Same check for empty `operatorIds` in `removeOperatorsWhitelistingContract`, `setOperatorsPrivateUnchecked`, `setOperatorsPublicUnchecked `

It's true that in `setOperatorsWhitelistingContract`, `removeOperatorsWhitelistingContract`, `setOperatorsPrivateUnchecked`, and `setOperatorsPublicUnchecked` there is no check for duplicate `operatorIds`. This only implies extra gas costs for the users that pass unique lists and no risk for the ones who pass duplicates.

worth adding some tests on duplicates? I can do at the end of priorities

Reply:
I think its worth to add a duplicate test but not important